ebs-backup.man
.TH ebs-backup 1 2022-04-25 2.0.2-alpha.1

.SH NAME
ebs-backup \- backup and restoration based on AWS EBS

.SH SYNOPSIS

.B ebs-backup b
.I keyname
.I zone
.IR file ...

.B ebs-backup r
.I keyname
.I zone

.SH DESCRIPTION
.B ebs-backup
uses the AWS key pair
.I keyname
to backup each
.I file
to an EBS volume in the availability zone named
.IR zone .
.B ebs-backup
can later restore each file using the same AWS key pair and availability zone.

The user MUST run the script once in backup mode (
.B ebs-backup b...
) at least once to setup a backup volume. This means that the user must have at
least one
.I file
available for backup before initial setup.

In any mode of operation (backup/restore),  if the user is not using an SSH
agent to cache their password, then the user must be present during the entire
runtime of the script to enter their SSH password. For more convenient use,
please use an SSH agent (see
.BR ssh-agent (1)).

.SS Limitations
The user CANNOT backup data with a total size (in 4096-byte-blocks) bigger than
the backup volume size (in bytes) that the user specified when they first ran
.B ebs-backup b...

The user CANNOT backup more than 16 TiB of data.

.SS Usage:
.B ebs-backup b
.I keyname
.I zone
.IR file ...

.RS 4
Will backup
.IR file (s)
using the AWS keypair 
.I keyname
and availability zone
.IR zone .
.RE

.B ebs-backup r
.I keyname
.I zone

.RS 4
Will restore the last backup. Files will be restored to their original state
and locations on the local machine.

.SS Important info

If the user has not created an AWS key pair, then the user must
create an AWS key pair (see
.UR https://\:docs.aws.amazon.com/\:AWSEC2/\:latest/\:UserGuide/\:ec2-key-pairs.html
.UE ).

Backup volumes are EBS volumes tagged with the Key "ebs_backup_vol" and Value
"1". If no EBS volumes have this key,value pair, then the user will be prompted
to create such a volume.
.B ebs-backup
will not perform backup or restoration unless a backup volume exists.

The provided
.I zone
must be the same as existing backup volumes, otherwise,
.B ebs-backup
cannot use the existing backup volumes and will run as if those volumes do not
exist.

Backups are located in a btrfs filesystem under "backup.n/". The smaller n is,
the more recent the backup (with n=0 indicating that the backup is the latest
backup).

.SS Rsync failures

If rsync fails, the script will usually just continue execution, terminating the
AWS EC2 instance used. Therefore, it is possible that that the latest backup
will be incomplete. When theis occurs, there is no way to prevent overwriting
backup.1 when
.B ebs-backup b...
runs again.

.SS Manual restoration

In the event of unforeseen cirumstances or if you need to restore a backup
earlier than the latest one, then you might need to perform manual restoration.

The paths of files relative to root (/) on the local system (where
.B ebs-backup
runs) are the same as the backed up files' paths relative to their respective
backup directory. Therefore, to restore the nth backup (starting at n=0), find
and attach the EBS volume tagged "ebs_backup_vol" to an EC2 instance and run:

.RS 4
$
.B scp -r
.BI ubuntu@ remote :mnt/backup.n/* 
/
.RE 

where remote is the public dns name of the EC2 instance in use.

.SH FILES
.I /usr/bin/ebs-backup
.RS 4
Primary script
.RE

.I /usr/share/man/man1/ebs-backup.1.gz
.RS 4
Man pages
.RE


ebs-backup.sh
#!/bin/sh

# Set program name now in case params shift or something
program="$0"
mode="$1"
key="$2"
zone="$3"

# Rationale for functions in here rather than a
# separate file:
# - User can just put this script anywhere
# - Minimize chance of breakage (no other files
#   that need to be anywhere specific)

###
# Error Handling
###

alias error='echo >&2'

fail() {
  error "$*"
  exit 1
}

usage() {
  echo "usage: $program b keyname zone file..."
  echo "       $program r keyname zone"
  exit 1
}

###
# AWS Instance Handling
###

# SSH without host key checking or saving the new host
alias ssh-tmp='ssh -q \
  -o StrictHostKeyChecking=no \
  -o UserKnownHostsFile=/dev/null \
  -o ConnectTimeout=15'

# Commands to issue over ssh (mainly mkfs and disk mounting.
# error() not defined on instance.
ssh_commands='echo "Making mountpoint..."
if ! mkdir mnt; then
  echo "Failed to make mountpoint" >&2
  exit 2
fi
sudo chown ubuntu:ubuntu mnt
if ! sudo mount /dev/xvdf mnt; then
  echo "Failed to mount backup volume"
  exit 3
fi'

terminate() {
  echo "Terminating instance..."
  aws ec2 terminate-instances \
    --instance-ids "$instance" >/dev/null
}

# "r" instead of "a" option due to errors with home directory perms:
# Can't change times
restore() {
  rsync -crvzP \
    -e 'ssh -q -o StrictHostKeyChecking=no
      -o UserKnownHostsFile=/dev/null
      -o ConnectTimeout=15' \
    "ubuntu@$iname:/home/ubuntu/mnt/backup.0/" /
  terminate
  exit 0
}

###
# Check if necessary commands exist
###

if ! command -v rsync >/dev/null; then
  error "Please install rsync(1)"
  exit 1
fi

# jq is used to parse aws output
if ! command -v jq >/dev/null; then
  error "Please install jq(1)"
  exit 1
fi

# TODO: Maybe add a check for aws, even if that's assumed to be installed?

###
# Check for parameters
###

# TODO: Assume default region is in config if not specified
# Rationale for forcing user to give zone name:
# - User zone *may* change between backups

# If backing up, expect params b, keyname, zone and file:
if [ "$mode" = "b" ] && [ $# -ge 4 ]; then
  shift 3
# If restoring, expect params r, keyname, zone:
elif [ "$mode" = "r" ] && [ $# = 3 ]; then
  :
else
  usage
  exit 1
fi

###
# Check availability zone
###

# List of available zones
zone_info=$(aws ec2 describe-availability-zones \
  | jq -r '.AvailabilityZones[] | select(.State == "available") | .ZoneName')

# printf for portability in unforeseen cases
if ! printf "%s" "$zone_info" | grep -qFx "$zone"; then
  error "Given zone not available."
  error "Available zones:"
  fail "$zone_info"
fi

###
# Check if backup volume(s) exist(s)
###

echo "Checking for backup volume(s)..."

ebs_vol=$(aws ec2 describe-volumes \
  | jq -r \
    --arg zone "$zone" \
    '.Volumes[]
    | select(.AvailabilityZone == $zone and .Tags[].Key == "ebs_backup_vol")
    | .VolumeId')

if [ -z "$ebs_vol" ]; then
  echo "Backup volume(s) not detected!"
  printf "Do you want to create a new volume? [default: n] [y/n]: "
  read -r ans
  ans=${ans:-n}
  if [ "$ans" != "y" ]; then
    error "No backup volume available."
    fail "See \`man 1 ebs-backup\`"
  fi
  

  printf "What size volume do you want (in GiB)? [default: 1] [1-16384]: "
  read -r size
  size=${size:-1}
  if ! ebs_vol=$(aws ec2 create-volume \
    --availability-zone "$zone" \
    --size "$size" \
    --volume-type gp3 \
    --tag-specifications 'ResourceType=volume,Tags=[{Key=ebs_backup_vol,Value="1"}]' \
    | jq -r '.VolumeId'); then
    fail "Volume creation failed."
  fi

  ssh_commands="sudo mkfs.btrfs /dev/xvdf; $ssh_commands"
fi

###
# Start an ephemeral ec2 instance
###

# Credit: Jan Schaumann <jschauma@netmeister.org>
# (Modified a bit from source)
# Start a fedora instance IPv4/IPv6 dual stack
# Assumes configuration as descrived in:
# https://www.netmeister.org/blog/ec2-ipv6.html
# Since the user set a default key-pair, we will
# not specify a key-pair for the run-instnaces
# command.
echo "Starting a temporary EC2 instance..."
subnet=$(aws ec2 describe-subnets \
  | jq -r '.Subnets[] | select( .Tags[]? | select(.Value == "dualstack")).SubnetId')
sg=$(aws ec2 describe-security-groups \
  | jq -r '.SecurityGroups [] | select( .GroupName == "dualstack").GroupId')
# Login with user "ubuntu" 
# Switch to an ami with a smaller ebs volume later.
if ! startup_json=$(aws ec2 run-instances --key-name "$key" \
  --placement AvailabilityZone="$zone" \
  --image-id ami-0f593aebffc0070e1 \
  --instance-type t2.micro  \
  --subnet-id "${subnet}"   \
  --security-group-ids "${sg}"); then
  fail "Failed to create instnace!"
fi

# Id of the temporary instance to perform backup to
instance=$(jq -nr --argjson data "$startup_json" '$data.Instances[].InstanceId')

# Wait for ec2 instance to come up
echo "Waiting for EC2 instance \"$instance\" to start running..."
aws ec2 wait instance-running --instance-ids "$instance"

###
# Attach volume to ec2 instance
###

if ! aws ec2 attach-volume \
  --device "/dev/xvdf" \
  --instance-id "$instance" \
  --volume-id "$ebs_vol" >/dev/null 2>&1; then
  error "Failed to attach volume."
  terminate
  exit 1
fi

# SSH may take time to setup, so wait 30 seconds.
secs=30
while [ $secs -ge 0 ]; do
  # Erase previous line and move cursor to beginning of current line
  printf "Seconds til SSH is likely active: %d\033[0K\r" "$secs"
  sleep 1
  secs=$((secs-1))
done

# Newline, since timer reset cursor position to start of line with text
echo ""
echo "Testing SSH..."

# Instance public dns name
iname=$(aws ec2 describe-instances --instance-ids "$instance" \
  | jq -r ".Reservations[].Instances[].PublicDnsName")


# Test ssh connection
if ! ssh-tmp "ubuntu@$iname" exit; then
  error "SSH connection took too long."
  terminate
  exit 1
fi
echo "Connection succeeded!"


if [ "$mode" = "b" ]; then
  ssh_commands="$ssh_commands; if [ ! -e mnt/backup.0 ]; then
  mkdir mnt/backup.0
  fi
  rm -rf mnt/backup.1
  mv mnt/backup.0 mnt/backup.1"
fi

ssh-tmp "ubuntu@$iname" "$ssh_commands"

###
# Restore if in restoration mode
###

if [ "$mode" = "r" ]; then
  restore
fi

###
# Making sure that backup can occur
###

echo "<--- Data transfer calculations --->"

# Compare filesize and file count with limits.
# Btrfs limits:
# - 2^64 files
# Note the 4096 blocksize

files=$(for file in "$@"; do
  printf "%s" "$(realpath "$file")"
  echo ""
done)

available=$(ssh-tmp "ubuntu@$iname" stat -f -c"%a" mnt)
files_used=$(ssh-tmp "ubuntu@$iname" "du -a mnt | wc -l")

du_out=$(du -caB 4096 "$files")
needed=$(printf "%s" "$du_out" | awk 'END{print $1}')

num_files=$(printf "%s" "$du_out" | head -n -1 | wc -l)
files_left=$(echo "18446744073709551616 - $num_files - $files_used" | bc)

printf "Needed blocks: %s\nAvailable blocks: %s\n" "$needed" "$available"
if [ "$(echo "$available - $needed > 0" | bc)" = "0" ]; then
  error "Not enough free blocks on backup volume!"
  terminate
  exit 1
fi

printf "Files to copy: %s\nNumber of files allowed: %s\n" "$num_files" "$files_left"
if [ "$(echo "18446744073709551616 - $num_files - $files_used" | bc)" = "0" ]; then
  error "Too many files will be backed up."
  terminate
  exit 1
fi

echo "<---------------------------------->"

###
# Copy files to backup
###

# -a: Archive mode ensures symbolic links, attributes,
#     perms, etc. are preserved.
# -v: Verbose
rsync -avRP --delete \
  -e 'ssh -q -o StrictHostKeyChecking=no
          -o UserKnownHostsFile=/dev/null
          -o ConnectTimeout=10' \
  --link-dest=../backup.1 \
  "$files" \
  "ubuntu@$iname:/home/ubuntu/mnt/backup.0"

###
# Terminate instnace
###

terminate
exit 0
